src/core/qgsmaprenderertask.cpp

```diff
@@ -190,6 +194,21 @@ bool QgsMapRendererTask::run()
       QRectF rect( 0, 0, img.width(), img.height() );
       pp.drawImage( rect, img, rect );
       pp.end();
+
+      if ( mSaveWorldFile )
+      {
+        CPLSetConfigOption( "GDAL_PDF_DPI", QString::number( mMapSettings.outputDpi() ).toLocal8Bit().constData() );
```

Comment:
You need to split this and save the QString::number( mMapSettings.outputDpi() ) result to a local variable before taking the data of it (otherwise it can point to a temporary location)

Reply:
There might be multithreading issues in using CPLSetConfigOption() because it affects the whole process. Better use CPLSetThreadLocalConfigOption() to restrict the effect to the current thread.This remark holds for most uses of CPLSetConfigOption() in QGIS where one sets it and unsets it a bit afterwards

Reply:
> You need to split this and save the QString::number( mMapSettings.outputDpi() ) result to a local variable before taking the data of it (otherwise it can point to a temporary location)

That's not true because here it is the argument of a function (that would be different if constData() would be affected to a variable, and then passed to CPLSetConfigOption()). The C++ standard guarantees that all temporary variables will be destroyed afterwards.
The output of the following snippet
```
#include <iostream>

struct X
{
    X() { std::cout << "X()" << std::endl; }
    ~X() { std::cout << "~X()" << std::endl; }
    int v() { return 0; }
};

void foo(int x)
{
    std::cout << "foo()" << std::endl;
}

int main()
{
    std::cout << "right" << std::endl;
    foo(X().v());

    std::cout << "wrong" << std::endl;
    int x = X().v();
    foo(x);
    return 0;
};

```
is
```
X()
foo()
~X()
wrong
X()
~X()
foo()
```

Reply:
@rouault I guess this does not hold true if the returned value is a pointer, correct?
Specifically clarifying with [#31144](https://github.com/qgis/QGIS/pull/31144/files) in mind.

```
feature.geometry().constGet()->centroid();
```